### PR TITLE
Application Case Updates - parallel tacsAIMs, composite materials, early_stopping, aero loads file fix, parallel meshing on analysis form of FuntofemShapeDriver

### PR DESCRIPTION
**Pull request by sean-engelstad**
* **Important:** parallel tacsAIMs for Nx speedup of tacsAIM.post_analysis() when there are N shape variables. Able to get ESP/CAPS runtime down from 40 mins/optimization iteration to only 12 mins/iteration on a large vehicle case.
* early_stopping is implemented but not tested
* composite materials for stringers
* output files for function values, derivatives and timing
* fixes for mesh morphing and remeshing f2f shape drivers
* improvements to aero_loads_file with `OnewayStructDriver` class - made the algorithm for reading in the aero loads more scalable aka faster. Also, improved the aero_loads_file unittest and fixed a bug where only the aero load for one index was accepted (now it works and matches the original aero loads).
* Demonstrated an internal structrure optimization using the `OnewayStructDriver` with an aero_loads_file.
* Add isnan checks to optimizer through the `OptimizationManager` class (experienced a nan ksfailure value in TACS which led to a failed ESP/CAPS build as the optimizer was not told that the analysis failed). This happened 20 optimization steps into my internal structure optimization with fixed aero loads and the `OnewayStructDriver` class.
* Improvements to the optimization plots (optional plot_name arg to function so the full_name doesn't go past the end of legend). Also, the x-axis for optimization history is an integer now.
* **Important:** add the ability for the using parallel tacsAIMs in the remeshing case by doing the mesh generation in the analysis script / driver (previously meshing done from the remeshing form of the driver). This required some changes to the `FuntofemShapeDriver`.

Note : this PR will probably not pass the unittests until the parallel tacsAIM / caps2tacs PR goes through.

### CODE REVIEW

**Comment by sean-engelstad** [comment]
Note the tests are failing currently since this depends on the parallel tacsAIM PR in TACS first.

**Review by bburke38 — CHANGES_REQUESTED** [review]
I'm going to pull this and make changes locally, then commit.

File: funtofem/optimization/openmdao_component.py [hunk 1]
```diff
@@ -139,12 +144,11 @@ def compute(self, inputs, outputs):
         track_history = self.options["track_history"]
         model = driver.model
         design_out_file = self.options["design_out_file"]
-        new_design = self.update_design(inputs, analysis=True)
+        self.update_design(inputs, analysis=True)
 
-        if new_design:
```
Comment (bburke38) [inline funtofem/optimization/openmdao_component.py:144]: Why aren't we still checking if it's a new design?

File: funtofem/optimization/openmdao_component.py [hunk 2]
```diff
@@ -156,11 +160,10 @@ def compute(self, inputs, outputs):
     def compute_partials(self, inputs, partials):
         driver = self.options["driver"]
         model = driver.model
-        new_design = self.update_design(inputs, analysis=False)
+        self.update_design(inputs, analysis=False)
 
-        if new_design:
```
Comment (bburke38) [inline funtofem/optimization/openmdao_component.py:161]: Why aren't we still checking if it's a new design?

File: funtofem/optimization/optimization_manager.py [hunk 3]
```diff
@@ -24,6 +24,7 @@
 
 import os, numpy as np
 import matplotlib.pyplot as plt
+from matplotlib.ticker import MaxNLocator
```
Comment (bburke38) [inline funtofem/optimization/optimization_manager.py:27]: We could make this generalizable to be able to change the ticker locator.

File: tests/unit_tests/framework/test_piston_framework.py [hunk 4]
```diff
@@ -144,7 +144,7 @@ def test_coupled_derivatives(self):
         # Check whether to use the complex-step method or now
```
Comment (bburke38) [inline tests/unit_tests/framework/test_piston_framework.py:144]:
```suggestion
        # Check whether to use the complex-step method or not
```

File: tests/unit_tests/shape/test_tacs_shape_driver.py [hunk 5]
```diff
@@ -188,6 +197,11 @@ def test_shape_and_thick_steady_aeroelastic(self):
             complex_mode=False,
             epsilon=1e-4,
         )
+
+        # for func in f2f_model.get_functions():
```
Comment (bburke38) [inline tests/unit_tests/shape/test_tacs_shape_driver.py:201]: If this is just commented out, can we get rid of it? Same with some previous lines in this file.

File: funtofem/optimization/openmdao_component.py (hunk 2 above)
Comment (sean-engelstad) [reply funtofem/optimization/openmdao_component.py:161]: I was having a problem getting it to run with the new_design mode. I will add an issue for this and address it later.

File: funtofem/optimization/optimization_manager.py (hunk 3 above)
Comment (sean-engelstad) [reply funtofem/optimization/optimization_manager.py:27]: maybe

File: tests/unit_tests/shape/test_tacs_shape_driver.py (hunk 5 above)
Comment (sean-engelstad) [reply tests/unit_tests/shape/test_tacs_shape_driver.py]: I removed the active=True arguments like you wanted. Also, removed the functions for loop at the bottom which I temporarily added for debugging this test with the new parallel tacsAIMs. I don't think we need to get rid of the TacsSteadyInterface line necessarily. It shows somebody how you could add that in in the parallel tacsAIM case if you want to.

File: funtofem/optimization/openmdao_component.py (hunk 1 above)
Comment (sean-engelstad) [reply funtofem/optimization/openmdao_component.py:144]: See below comment, I will make an issue for this.

File: funtofem/optimization/optimization_manager.py (hunk 3 above)
Comment (bburke38) [reply funtofem/optimization/optimization_manager.py:27]: Not saying right now, but in the future.

**Review by bburke38 — APPROVED** [review]

**Pull request merged**